include/rmm/cuda_stream_pool.hpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream.hpp>
+#include <rmm/cuda_stream_view.hpp>
+
+#include <atomic>
+#include <vector>
+
+namespace rmm {
+
+/**
+ * @brief A pool of CUDA streams.
+ *
+ * Provides efficient access to collection of CUDA stream objects.
+ *
+ * Successive calls may return a `cuda_stream_view` of identical streams. For example, a possible
```

Comment:
This makes this pool unusable for DALI - unwittingly submitting new (and potentially urgent) work on a stream that already has hundreds of milliseconds worth of work scheduled is not an option.

Reply:
You do know that there are hardware limits to concurrency, right? See table 15 in the CUDDA programming guide. The current high end GPUs have at most 128 concurrent execution contexts, so using more than 128 streams is redundant -- you will get false dependencies even with more streams.

The size of the pool is configurable. Set it to 128 if you want 128 streams.

We can also discuss what the best default size is before we merge this. I chose 16 because that is the minimum maximum concurrency of GPUs that RAPIDS supports today. I didn't want to create hundreds of streams if the user doesn't need them.

We can also consider a pool that can grow, with a parameterized maximum size.

Reply:
As we make the pool more complex, we can also have a notion of normal streams and high priority streams, e.g.,  `get_stream` and `get_high_priority_stream`.

Reply:
I guess that we have a very different idea of a stream pool then.

My idea (which I'm going to pursue in DALI) was to have an pool where you take streams from the pool and explicitly return them. That's more or less equivalent to create/destroy, but without the ugly handle reuse.

The pool would have two lists:

- ready list
- free list

When a client requests a stream, it's taken from the ready list. When it's returned to the pool, it's placed in the free list. Then there are several options:
1. Add a callback to the stream, so that it's returned to the ready list when all work is complete.
2. Call cudaStreamQuery periodically in a worker thread and collect ready streams.
3. Call cudaStreamQuery on the streams in the free list when there's no ready stream and return the ready ones to the ready list (this is the least preferred option, as it introduces a potentially large and unpredictable overhead).

In any case, if there are no ready streams and there are no ready streams, a new stream is created and returned.

Reply:
I did consider a design with get_stream and return_stream, but I think this is early optimization. The present design is extremely simple, and will likely provide good performance for many use case. Once we find that the performance is insufficient, we can add complexity as needed. Certainly, a design that requires worker threads is much more complexity, and I will not add that complexity to RMM without a real-world benchmark to drive it.